internal/pkg/watch/registry.go

```diff
@@ -0,0 +1,122 @@
+package watch
+
+import (
+	"sync"
+)
+
+// DefaultWatchRegistry is the default watcher registry used by the agent.
+var DefaultWatchRegistry WatchersRegisterer
+
+func init() {
+	defaultWatchRegistrar := new(Registry)
+
+	defaultWatchRegistrar.watch = []*WatcherInstance{}
+	defaultWatchRegistrar.Mutex = &sync.Mutex{}
+	DefaultWatchRegistry = defaultWatchRegistrar
+}
+
+// WatchersRegisterer is an interface for enabling agent watchers.
+type WatchersRegisterer interface {
+	Register(w ...Watcher) error
+	Start(ch ...chan<- interface{}) error
+	RegisterAndStart(w Watcher, ch ...chan<- interface{}) error
+	Stop()
+	Wait()
+}
+
+// Registry type
+type Registry struct {
+	watch []*WatcherInstance
+	*sync.Mutex
+}
+
+// WatcherInstance describes a state of a single
+// watcher that's inside the registry.
+type WatcherInstance struct {
+	started bool
+	watcher Watcher
+	*sync.Mutex
+}
+
+// Register registrers one or more watchers
+func (r *Registry) Register(w ...Watcher) error {
+	r.Lock()
+	defer r.Unlock()
+	for _, watcher := range w {
+		_, err := r.register(watcher)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (r *Registry) register(watcher Watcher) (*WatcherInstance, error) {
+	instance := &WatcherInstance{
+		watcher: watcher,
+		Mutex:   &sync.Mutex{},
+	}
+	r.watch = append(r.watch, instance)
+	return instance, nil
+}
+
+// RegisterAndStart attempts to register and start a single watcher.
+func (r *Registry) RegisterAndStart(w Watcher, ch ...chan<- interface{}) error {
```

Comment:
We need to make sure this function is idempontent so that a watcher instance cannot be registered nor started multiple times. The goroutine that ticks and starts the watcher when node role is discovered, has a bug (missing `return`) and calls `RegisterAndStart()` multiple times, which results in multiple log watchers being spawned ending up producing so many data that eventually get dropped by the exporter channel emit system.    

I think you need to use a similar interface for `Registry` as we do for `Watch`, which uses a `sync.Once` instance to implement this safety.

Reply:
Addressed in [75549ec](https://github.com/Metrika-Inc/agent/pull/95/commits/75549ec1d6e26605c6106991a7494aa244a643fd)